docs/dev/replay-exec.md

```diff
@@ -0,0 +1,86 @@
+This doc describes how to dump an Exec runtime meta and column batch data to a directory,
+and replay the Exec with the dumped meta and data. When encountering a perf issue using this tool
+can dump the runtime meta/data and then replay. It will spend less time when collecting NSYS and
+NCU information when replaying the dumped runtime meta/data.
+
+# Dump a Exec meta and a column batch data
+
+## compile Spark-Rapids jar
+e.g.: compile for Spark330
+```
+mvn clean install -DskipTests -pl dist -am -DallowConventionalDistJar=true -Dbuildver=330 
+```
+Note: Should specify `-DallowConventionalDistJar`, this config will make sure to generate a
+conventional jar. If you do not specify this config, then replay will fail because of can not
+find the class `com.nvidia.spark.rapids.debug.ProjectExecReplayer`
+
+## enable dump
+This assumes that the RAPIDs Accelerator has already been enabled.   
+This feature is disabled by default.   
+Set the following configurations to enable this feature:
+
+``` 
+spark.conf.set("spark.rapids.sql.debug.replay.exec.types", "project")
+```
+Default `types` value is empty which means do not dump.   
+Define the Exec types for dumping, separated by comma, e.g.: `project,aggregate,sort`.   
+Note currently only support `project`
+
+```
+spark.conf.set("spark.rapids.sql.debug.replay.exec.dumpDir", "file:/tmp")
+```
+Specify the dump directory which can be a local path or a remote path. E.g.: 
+`file:/tmp/my-debug-path`. Default value is `file:/tmp`.
+This path should be a directory or someplace that we can create a directory to
+store files in. Remote path is supported e.g.: `hdfs://url:9000/path/to/save`. When using
+remote path, should specify corresponding configs to get access for the remote storage.
+
+```
+spark.conf.set("spark.rapids.sql.debug.replay.exec.threshold.timeMS", 1000)
+```
+Default value is 1000MS.   
+Only dump the column batches when it's executing time exceeds threshold time.
+
+```  
+spark.conf.set("spark.rapids.sql.debug.replay.exec.batch.limit", 1)
+```
+This config defines the max dumping number of column batches.   
+Default value is 1.
+
+## run a Spark job
+This dumping only happens when GPU Exec is running, so should enable Spark-Rapids.
+After the job is done, check the dump path will have files like:
+```
+/tmp/replay-exec:
+  - xxx_GpuTieredProject.meta      // this is serialized GPU Tiered Project case class  
```

Comment:
if I have a huge job that has say 10000 projects is this directory going to become to big?  I assume the job keeps running after dumping this out, correct?  Is there a way to limit the total number you get?

Reply:
> if I have a huge job that has say 10000 projects is this directory going to become to big? 

Yes, it will be big.  We have a config `threshold.timeMS` to reduce the number of batches to dump.
Typically usage is: Running the query at customer env, then check the eventlog to find a threshold time; then run at customer env again with dump enabled and with this threshold time. Usually we only need one batch to reproduce the slowness, and reproduce with NSYS.

> I assume the job keeps running after dumping this out, correct? 

Yes. Keep runnig, if it's needed we can terminate the process when the first dumping is done.

> Is there a way to limit the total number you get?

No. In future will collect batches for multiple Execs. We will do not know how many     slowness batches in advance.